### PR TITLE
Document ID and EUI constraints

### PR DESCRIPTION
**Pull request by benolayinka**
#### Summary

Closes https://github.com/TheThingsIndustries/lorawan-stack-docs/issues/211
Closes https://github.com/TheThingsIndustries/lorawan-stack-support/issues/267

#### Changes

- Document how TTS uses IDs and EUIs

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Run Locally: Verified that the docs build using `make server`, posted screenshots, verified external links.
- [ ] New Features Marked: Documentation for new features is marked using the `new-in-version` shortcode, according to the guidelines in [CONTRIBUTING](CONTRIBUTING.md).
- [ ] Style Guidelines: Documentation obeys style guidelines in [CONTRIBUTING](CONTRIBUTING.md).
- [x] Commits: Commit messages follow guidelines in [CONTRIBUTING](CONTRIBUTING.md), there are no fixup commits left.

### CODE REVIEW

File: doc/content/reference/id-eui-constraints/index.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+This reference explains how {{% tts %}} deals with IDs and EUIs, including regular expression requirements, uniqueness, and how deletion is handled.
+
+<!--more-->
+
+## Requirements of an ID or EUI
+
+An ID or EUI in {{% tts %}} must:
+
+- Have a length of 3-36 characters
```
Comment (adriansmares) [inline doc/content/reference/id-eui-constraints/index.md:14]: I'd say between 3 and 36 characters.

File: doc/content/reference/id-eui-constraints/index.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+- Consist of lowercase letters, numbers, and non-consecutive dashes
```
Comment (adriansmares) [inline doc/content/reference/id-eui-constraints/index.md:15]: You cannot start with a dash either, but I don't know how we can spell this out in plain English. Maybe a 3rd rule that says 'must not start with a dash' ?

File: doc/content/reference/id-eui-constraints/index.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+`(^[a-z0-9](?:[-]?[a-z0-9]){2,}$)`
+
+## Uniqueness
+
+Within a deployment, there can only be:
+
+- One tenant with the same ID
+- One gateway with the same EUI (across all tenants)
+- One end device with the same DevEUI (across all tenants)
```
Comment (adriansmares) [inline doc/content/reference/id-eui-constraints/index.md:27]: This is not true. A DevEUI may be used in two tenants.

File: doc/content/reference/id-eui-constraints/index.md (hunk 1 above)
Comment (htdvisser) [reply doc/content/reference/id-eui-constraints/index.md:14]: FYI: There is also an open issue about lowering the minimum from 3 to 2: https://github.com/TheThingsNetwork/lorawan-stack/issues/3723

File: doc/content/reference/id-eui-constraints/index.md (hunk 3 above)
Comment (benolayinka) [reply doc/content/reference/id-eui-constraints/index.md:27]: took this from @htdvisser comment https://github.com/TheThingsIndustries/lorawan-stack-docs/issues/211#issue-795879347.

mistake?

Comment (htdvisser) [reply doc/content/reference/id-eui-constraints/index.md:28]: According to the definition of a Dev EUI, it is a globally unique identifier, so across all tenants.

The Identity Server only enforces uniqueness of the (JoinEUI,DevEUI) tuple across all tenants. For the record: I think we should be more strict, but this is what we decided to do.

Comment (benolayinka) [reply doc/content/reference/id-eui-constraints/index.md]: Updated, awaiting review @adriansmares

File: doc/content/reference/id-eui-constraints/index.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+`(^[a-z0-9](?:[-]?[a-z0-9]){2,}$)`
+
+## Uniqueness
+
+Within a deployment, there can only be:
+
+- One tenant with the same ID
+- One gateway with the same EUI (across all tenants)
+- One end device with the same DevEUI,JoinEUI tuple (across all tenants)
```
Comment (htdvisser) [inline doc/content/reference/id-eui-constraints/index.md:28]: Maybe "DevEUI,JoinEUI tuple" is too technical, you can also write "combination of JoinEUI and DevEUI"

File: doc/content/reference/id-eui-constraints/index.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+Within an application, there can only be:
+
+- One end device with the same ID
+
+## Deleted Entities
+
+When deleting entities, some IDs are retained so that they can't be registered again. This is for security reasons as integrations may store historical data by ID, and you don't want that someone else can register an ID and get your historical data.
+
+EUIs are released in some cases so that they can be registered in other applications or tenants.
```
Comment (htdvisser) [inline doc/content/reference/id-eui-constraints/index.md:45]: The EUIs of Gateways and End Devices are released when they are deleted.

File: doc/content/reference/id-eui-constraints/index.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+- When a user or organization is deleted, the ID is retained and cannot be reused within the deployment
```
Comment (htdvisser) [inline doc/content/reference/id-eui-constraints/index.md:47]: I think we already covered the "within the xxx" above. Otherwise this is incorrect, because the same user/org IDs can be used in different tenants.

Comment (htdvisser) [reply doc/content/reference/id-eui-constraints/index.md:47]: Same story below.

**Review by adriansmares — APPROVED** [review]
My comments are addressed, thanks. Please address the ones from @htdvisser.

File: doc/content/reference/id-eui-constraints/index.md (hunk 6 above)
Comment (benolayinka) [reply doc/content/reference/id-eui-constraints/index.md]: Okay. I'm going to leave the redundant but factually correct statements just to reiterate that the deleted IDs are retained and can't be used in the tenant, if that's correct:

- When a user or organization is deleted, the ID is retained and cannot be reused within the tenant
- When an application is deleted, the ID is retained and cannot be reused within the tenant
- When a gateway is deleted, the ID is retained but the gateway EUI can be reused
- When an end device is deleted, the ID and EUI are released and can be reused

**Review by htdvisser — APPROVED** [review]
